### DISCUSSION TITLE
Question regarding the test-timing code

### DISCUSSION THREAD

**Discussion by efthym** [jefftenney/LPTIM-Tick#16] [Q&A] [answered]
Hi Jeff,
Thanks for sharing LPTIM-Tick. I'm using it to get a better understanding of FreeRTOS tickless mode using an STM32H7 on a Nucleo board.
I'm thinking of employing your method for testing the timing accuracy using RTC to other tickless implementations, such use the default configUSE_TICKLESS_IDLE = 1 or a method in an STM32 MOOC (https://www.youtube.com/playlist?list=PLnMKNibPkDnExrAsDpjjF1PsvtoAIBquX  video #57 onwards).

In your tick-timing task vTttOsTask, you have a loop waiting for the tickCount to change:
```
         do {
            xTickCount = xTaskGetTickCount();
            rtcTimeAtTick = tickHookSnapshot; // structure copy

         } while (xTickCount != xTaskGetTickCount());
```
I understand that in order to exit the loop,  you must get interrupted, by the tick ISR, between the two calls to xTaskGetTickCount(). However this can happen either before the tickHookSnaphot is copied or after. In the first case you get the new RTC timestamp of the tick and in the second you get the old timestamp. You might even get the interrupt in the middle of copying the structure.
Wouldn't this matter in your subsequent calculations?
Why don't you use the value of tickHookSnaphot just after the loop, which should be the "new" timestamp? (Assuming vTttOsTask has the highest priority, nothing can run between the exit from the while and the copy of the timestamp)

Thanks,
Aris

**Comment by jefftenney** [comment, marked as answer]
Hi @efthym -

That loop might look like it is waiting for the tickCount to change, but really it's the opposite.  The loop only iterates again if the tick count changes.  So in reality, the loop almost never iterates.  It is capturing two items that can change asynchronously (the tick-hook RTC snapshot and the OS tick count) and then verifying afterward that neither one changed during/after the capturing.  Thus the two items can be considered coherent (or correlated correctly to eachother).  Does that help?

**Reply by efthym** [reply]
Yes it does help, thanks a lot. My mind got stuck and I didn't see what was there!

**Comment by jefftenney** [comment]
@efthym The Compare-ST-Example branch might be helpful if you are evaluating ST's code for LPTIM FreeRTOS tick.  Take a look.
